lib/hypothesis/engine.rb

```diff
@@ -7,27 +7,35 @@ module Hypothesis
   class Engine
     attr_reader :current_source
 
-    def initialize(options)
-      @max_examples = options.fetch(:max_examples, 200)
-      @core_engine = HypothesisCoreEngine.new(
-        options[:seed] || Random.rand(2**64 - 1)
-      )
+    def initialize(max_examples: 200, seed: nil)
+      if seed.nil?
+        seed = Random.rand(2**64 - 1)
+      end
+      @core_engine = HypothesisCoreEngine.new(seed, max_examples)
     end
 
     def run
-      count = 0
-      total_count = 0
-      while (count < @max_examples) && (total_count < @max_examples * 10)
-        @current_source = Source.new(@core_engine, @core_engine.new_source)
-        count += 1
-        total_count += 1
+      while @core_engine.should_continue
+        core_id = @core_engine.new_source
+        @current_source = Source.new(@core_engine, core_id)
         begin
           yield(@current_source)
         rescue UnsatisfiedAssumption
-          count -= 1
+          @core_engine.finish_invalid(core_id)
+        rescue DataOverflow
+          @core_engine.finish_overflow(core_id)
+        rescue Exception
+          @core_engine.finish_interesting(core_id)
+        else
+          @core_engine.finish_valid(core_id)
```

Comment:
Is there a reason you used the `else` here, instead of putting finish_valid after the yield (which is probably more Ruby)